[test] Add test for gettxout to wallet.py

jimmysong opened the pull request:
Test gettxout as part of the wallet test.
Sanity-checks the utxo generated from coinbase.

TheBlueMatt commented:
utACK. It would be cool to also test the include_mempool flag to gettxout to check that a) things which are spent in the mempool arent visible via gettxout (even though they are without include_mempool) and b) things which are only available in the mempool are inversely visible.

jimmysong committed:
[test] Add gettxout call

Test gettxout as part of the wallet test.
Tests gettxout with a confirmed/unconfirmed tx with include_mempool flag on and off

jimmysong commented:
@TheBlueMatt Thank you for your excellent suggestion. I've made the changes as you suggested, squashed and rebased.

jnewbery commented on test/functional/wallet.py:
```diff
@@ -52,13 +52,33 @@ def run_test (self):
-        self.nodes[0].sendtoaddress(self.nodes[2].getnewaddress(), 10)
+        mempool_txid = self.nodes[0].sendtoaddress(self.nodes[2].getnewaddress(), 10)
+
+        # test gettxout
```
nit: my personal preference is to make this kind of 'heading' comment into a log, so test runners have some feedback as they're running the test.

jnewbery commented on test/functional/wallet.py:
```diff
@@ -52,13 +52,33 @@ def run_test (self):
+        mempool_txid = self.nodes[0].sendtoaddress(self.nodes[2].getnewaddress(), 10)
+
+        # test gettxout
+        # confirmed tx should be visible if you exclude mempool
```
This comment isn't quite right. We're not looking at a confirmed transaction here, we're looking at a transaction output who's spend is unconfirmed. That's why if include_mempool is set to False, we still think we own this output

jnewbery commented on test/functional/wallet.py:
```diff
@@ -52,13 +52,33 @@ def run_test (self):
+        txout = self.nodes[0].gettxout(confirmed_txid, confirmed_index, False)
+        assert_equal(txout['value'], 50)
+        txout = self.nodes[0].gettxout(confirmed_txid, confirmed_index, True)
+        assert(txout is None)
```
no need for brackets here. [assert is a statement, not a function](http://stackoverflow.com/a/30770447/933705).

This should just be: `assert not txout`

jnewbery commented on test/functional/wallet.py:
```diff
@@ -52,13 +52,33 @@ def run_test (self):
+        # note the mempool tx will have randomly assigned indices
+        # but 10 will go to node2 and the rest will go to node0
+        balance = self.nodes[0].getbalance()
+        assert(txout['value'] in (10, balance))
```
I think the intent is clearer if you collect both txouts above, and then test that one of them is equal to 10 and the other is equal to balance. Something like:

```
txout0 = self.nodes[0].gettxout(mempool_txid, 0, True)
txout1 = self.nodes[0].gettxout(mempool_txid, 1, True)

balance = self.nodes[0].getbalance()
assert_equal(set(10, balance), set(txout0['value'], txout1['value']))
```

jimmysong replied:
good idea

jnewbery commented on test/functional/wallet.py:
```diff
@@ -52,13 +52,33 @@ def run_test (self):
+        # confirmed tx should be visible if you exclude mempool
+        # but invisible if you include mempool
+        confirmed_txid, confirmed_index = utxos[0]["txid"], utxos[0]["vout"]
+        txout = self.nodes[0].gettxout(confirmed_txid, confirmed_index, False)
```
consider using named arguments when the positional arguments are just bools. It makes it difficult for the reader to know what the argument is for (alright, I know you've explained it in the comment above, but the reader still needs to trust that you're right!)

jnewbery reviewed (commented):
Great. Thanks for adding this.

A few nits, but otherwise looks good.

jimmysong commented:
nits addressed, rebased and squashed.

MarcoFalke commented on test/functional/wallet.py:
```diff
@@ -52,13 +52,33 @@ def run_test (self):
+        txout = self.nodes[0].gettxout(mempool_txid, 0, False)
+        assert txout is None
+        txout1 = self.nodes[0].gettxout(mempool_txid, 0, True)
+        txout2 = self.nodes[0].gettxout(mempool_txid, 1, True)
```
nit: In the future you might use named args for arguments that are unnamed boolean or integers of a function that takes more than one arg.
